e2e/e2e_issue_test.go

```diff
@@ -0,0 +1,94 @@
+package e2e
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/0xPolygon/pbft-consensus"
+)
+
+func TestE2E_Partition_MinorityCantValidate(t *testing.T) {
+	const nodesCnt = 7 // N=3F + 1, F = 2
+	hook := newPartitionTransport(300 * time.Millisecond)
+
+	c := newPBFTCluster(t, "majority_partition", "prt", nodesCnt, hook)
+	limit := int(math.Floor(nodesCnt*2.0/3.0)) + 1 // 2F+1 nodes can Validate
+	for _, node := range c.Nodes() {
+		node.StartWithBackendFactory(createBackend(node.name >= "prt_"+strconv.Itoa(limit)))
+	}
+	defer c.Stop()
+	names := generateNodeNames(0, limit, "prt_")
+	err := c.WaitForHeight(4, 1*time.Minute, names)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestE2E_Partition_MajorityCantValidate(t *testing.T) {
+	const nodesCnt = 7 // N=3F + 1, F = 2
+	hook := newPartitionTransport(300 * time.Millisecond)
+
+	c := newPBFTCluster(t, "majority_partition", "prt", nodesCnt, hook)
+	limit := int(math.Floor(nodesCnt * 2.0 / 3.0)) // + 1 removed because 2F+1 nodes is majority
+	for _, node := range c.Nodes() {
+		node.StartWithBackendFactory(createBackend(node.name < "prt_"+strconv.Itoa(limit)))
+	}
+	defer c.Stop()
+	names := generateNodeNames(limit, nodesCnt, "prt_")
+	err := c.WaitForHeight(3, 1*time.Minute, names)
+	if err == nil {
+		t.Fatal("Height reached for minority of nodes")
+	}
+}
+
+func TestE2E_Partition_BigMajorityCantValidate(t *testing.T) {
+	const nodesCnt = 100
+	hook := newPartitionTransport(300 * time.Millisecond)
+
+	c := newPBFTCluster(t, "majority_partition", "prt", nodesCnt, hook)
+	limit := int(math.Floor(nodesCnt * 2.0 / 3.0)) // + 1 removed because 2F+1 nodes is majority
+	for _, node := range c.Nodes() {
+		node.StartWithBackendFactory(createBackend(node.name <= "prt_"+strconv.Itoa(limit)))
+	}
+	defer c.Stop()
+	nodeNames := generateNodeNames(limit, nodesCnt, "prt_")
+	err := c.WaitForHeight(8, 1*time.Minute, nodeNames)
+	if err == nil {
+		t.Fatal("Height reached for minority of nodes")
+	}
+}
+
+type fsmValidationError struct {
+	fsm
+}
+
+func (f *fsmValidationError) Validate(_ []byte) error {
+	return fmt.Errorf("invalid propsal")
+}
+
+type fsmInsertError struct {
+	fsm
+}
+
+func (f *fsmInsertError) Insert(_ *pbft.SealedProposal) error {
+	return fmt.Errorf("insert error")
+}
+
+func createBackend(invalid bool) func(nn *node) pbft.Backend {
+	return func(nn *node) pbft.Backend {
+		if invalid {
+			switch rand.Intn(2) {
```

Comment:
What is this? If this are not fuzz tests why do they have this random here?

Reply:
It just creates a backend with "bad" behaviour, node that always error on validation or node that always return error on insert. It just picks randomly one of these 2 implementations. We can use only one implementation and remove random if needed.